File: backend/src/app/access/store.py
"""Access accounting — who uses the dashboard, persistent and flood-proof.

ADR docs/access-log.md. Same two-tier shape as ``geoip/denials.py`` (DR-G9):

- ``access_stats`` — per (hour, principal) counter, upserted by the flush job.
  Counts EVERY request; cardinality is hours × active principals, so neither
  a polling UI nor a scanner can flood it.
- ``access_events`` — sampled per-request rows for the timeline, hard-capped
  per flush interval (``_EVENTS_PER_FLUSH``). Only user sessions are sampled:
  anon requests aggregate without IP rows (DR-AL8), API-key scrapes are
  uniform noise.
- ``auth_sessions`` — login-session bookkeeping (DR-AL3, NOT enforcement):
  created on login, ``last_seen_at`` stamped throttled via the flush job,
  expiry job turns the silent 12h cookie death into a real audit event
  (``auth.session_expired``, DR-AL4).

The request path only touches bounded in-memory buffers; a 15s scheduler job
flushes them with its own session. Process-local by design (single worker).
"""

from __future__ import annotations

import time
import uuid
from collections import Counter, deque
from datetime import UTC, datetime, timedelta

import structlog
from sqlalchemy import delete, select, text, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.audit.log import write_audit
from app.db.models import AccessEvent, AccessStat, AuthSession

log = structlog.get_logger("app.access")

_EVENTS_PER_FLUSH = 50  # row-sample cap per flush interval; aggregate counts all
_LAST_SEEN_THROTTLE_S = 60.0  # max one last_seen write per session per window
_PRUNE_BATCH = 10_000

# Keep in sync with SessionMiddleware max_age in main.py (imported there).
SESSION_MAX_AGE_S = 12 * 60 * 60
# A session counts as online while last_seen is younger than this — 5× the
# stamp throttle, so an active session can never flicker offline.
ONLINE_WINDOW_S = 5 * 60

# --- request-path buffers (bounded) -------------------------------------------
_pending_agg: Counter[tuple[str, str]] = Counter()  # (ptype, pkey) -> n
_pending_last_ip: dict[tuple[str, str], str] = {}
_pending_events: deque[dict] = deque(maxlen=_EVENTS_PER_FLUSH)
_pending_seen: dict[str, datetime] = {}  # sid -> last_seen to persist
_seen_stamped: dict[str, float] = {}  # sid -> monotonic of last stamp (throttle)


def new_sid() -> str:
    return uuid.uuid4().hex


def record_request(
    ptype: str,
    pkey: str,
    ip: str | None,
    method: str,
    path: str,
    status: int,
    *,
    user_id: int | None = None,
    sid: str | None = None,
) -> None:
    """Count one request; sample a detail row only for user sessions (DR-AL8)."""
    _pending_agg[(ptype, pkey)] += 1
    if ip:
        _pending_last_ip[(ptype, pkey)] = ip[:45]
    if ptype == "user":
        _pending_events.append(
            {
                "ts": datetime.now(UTC),
                "user_id": user_id,
                "ip": (ip or "?")[:45],
                "method": method[:8],
                "path": path[:255],
                "status": status,
            }
        )
    if sid:
        now = time.monotonic()
        last = _seen_stamped.get(sid)
        if last is None or now - last >= _LAST_SEEN_THROTTLE_S:
            _seen_stamped[sid] = now
            _pending_seen[sid] = datetime.now(UTC)


def record_apikey(key_id: int, ip: str | None) -> None:
    """API-key hit — aggregate only (scrapes are uniform, no sample rows)."""
    _pending_agg[("apikey", str(key_id))] += 1
    if ip:
        _pending_last_ip[("apikey", str(key_id))] = ip[:45]


async def flush(session: AsyncSession) -> int:
    """Write pending buffers to the DB. Returns the number of aggregated requests.

    Caller (scheduler job) owns the commit. Buffers are swapped out first so a
    failed commit loses at most one interval — acceptable for statistics.
    """
    if not _pending_agg and not _pending_events and not _pending_seen:
        return 0
    agg = dict(_pending_agg)
    _pending_agg.clear()
    last_ips = dict(_pending_last_ip)
    _pending_last_ip.clear()
    events = list(_pending_events)
    _pending_events.clear()
    seen = dict(_pending_seen)
    _pending_seen.clear()

    bucket = datetime.now(UTC).replace(minute=0, second=0, microsecond=0)
    for (ptype, pkey), n in agg.items():
        # MariaDB-native upsert (repo rule: no ON CONFLICT idioms).
        await session.execute(
            text(
                "INSERT INTO access_stats (bucket, principal_type, principal_key, count, last_ip)"
                " VALUES (:bucket, :ptype, :pkey, :n, :ip)"
                " ON DUPLICATE KEY UPDATE count = count + :n,"
                " last_ip = COALESCE(:ip, last_ip)"
            ),
            {
                "bucket": bucket,
                "ptype": ptype,
                "pkey": pkey,
                "n": n,
                "ip": last_ips.get((ptype, pkey)),
            },
        )
    for e in events:
        session.add(AccessEvent(**e))
    for sid, ts in seen.items():
        await session.execute(
            update(AuthSession)
            .where(AuthSession.sid == sid, AuthSession.ended_at.is_(None))
            .values(last_seen_at=ts)
        )
    await session.flush()
    return sum(agg.values())


async def open_session(session: AsyncSession, *, sid: str, user_id: int, ip: str | None) -> None:
    """Registry row for a freshly minted login session (caller commits)."""
    now = datetime.now(UTC)
    session.add(AuthSession(sid=sid, user_id=user_id, ip=ip, created_at=now, last_seen_at=now))
    _seen_stamped[sid] = time.monotonic()
    await session.flush()


async def close_session(session: AsyncSession, sid: str | None, reason: str) -> None:
    """Mark a registry row ended (logout). Missing/unknown sid is a no-op —
    sessions minted before migration 041 have no row."""
    if not sid:
        return
    await session.execute(
        update(AuthSession)
        .where(AuthSession.sid == sid, AuthSession.ended_at.is_(None))
        .values(ended_at=datetime.now(UTC), end_reason=reason)
    )
    _seen_stamped.pop(sid, None)
    _pending_seen.pop(sid, None)


async def expire_sessions(session: AsyncSession) -> int:
    """Mark sessions older than the cookie max_age as expired + audit each.

    This is bookkeeping, not enforcement (DR-AL3/AL4): the cookie died on its
    own at max_age — here the previously silent auto-logout becomes a visible
    ``auth.session_expired`` event. Caller commits.
    """
    cutoff = datetime.now(UTC) - timedelta(seconds=SESSION_MAX_AGE_S)
    rows = (
        (
            await session.execute(
                select(AuthSession).where(
                    AuthSession.ended_at.is_(None), AuthSession.created_at < cutoff
                )
            )
        )
        .scalars()
        .all()
    )
    now = datetime.now(UTC)
    for s in rows:
        s.ended_at = now
        s.end_reason = "expired"
        _seen_stamped.pop(s.sid, None)
        await write_audit(
            session,
            action="auth.session_expired",
            result="ok",
            user_id=s.user_id,
            source_ip=s.ip,
        )
    if rows:
        await session.flush()
    return len(rows)


async def prune(
    session: AsyncSession, *, events_days: int, stats_days: int, sessions_days: int
) -> None:
    """Retention: batched oldest-first deletes (unbounded DELETE gap-locks)."""
    event_cutoff = datetime.now(UTC) - timedelta(days=events_days)
    while True:
        result = await session.execute(
            delete(AccessEvent).where(
                AccessEvent.id.in_(
                    select(AccessEvent.id)
                    .where(AccessEvent.ts < event_cutoff)
                    .order_by(AccessEvent.ts)
                    .limit(_PRUNE_BATCH)
                    .scalar_subquery()
                )
            )
        )
        await session.commit()
        if (result.rowcount or 0) < _PRUNE_BATCH:
            break
    stats_cutoff = datetime.now(UTC) - timedelta(days=stats_days)
    await session.execute(delete(AccessStat).where(AccessStat.bucket < stats_cutoff))
    await session.commit()
    sessions_cutoff = datetime.now(UTC) - timedelta(days=sessions_days)
    await session.execute(
        delete(AuthSession).where(
            AuthSession.ended_at.is_not(None), AuthSession.ended_at < sessions_cutoff
        )
    )
    await session.commit()


# --- scheduler entry points ----------------------------------------------------


async def flush_job() -> None:
    """15s: flush counters, samples and last_seen stamps with an own session."""
    from app.db.base import get_sessionmaker  # local import: avoid cycles

    if not _pending_agg and not _pending_events and not _pending_seen:
        return
    async with get_sessionmaker()() as session:
        n = await flush(session)
        await session.commit()
    if n:
        log.debug("access.flushed", requests=n)


async def expire_job() -> None:
    """5 min: turn max_age session deaths into audit events."""
    from app.db.base import get_sessionmaker

    async with get_sessionmaker()() as session:
        n = await expire_sessions(session)
        await session.commit()
    if n:
        log.info("access.sessions_expired", sessions=n)


async def prune_job() -> None:
    """Daily: retention for samples, aggregates and ended sessions."""
    from app.db.base import get_sessionmaker
    from app.settings.store import effective_settings

    settings = effective_settings()
    async with get_sessionmaker()() as session:
        await prune(
            session,
            events_days=int(settings.access_events_retention_days),
            stats_days=int(settings.access_stats_retention_days),
            sessions_days=int(settings.access_sessions_retention_days),
        )

File: orbit/lib/orbit/access/store.ex
defmodule Orbit.Access.Store do
  @moduledoc """
  Access accounting — who uses the dashboard, persistent and flood-proof.
  Port of access/store.py (ADR docs/access-log.md), writing the SAME
  alembic-owned tables as the python stack (upserts additive, prune
  idempotent — both stacks can account concurrently).

  - `access_stats` — per (hour, principal) counter upsert; counts EVERY
    request, cardinality is hours × active principals (DR-AL2).
  - `access_events` — sampled per-request rows, hard-capped per flush
    interval. Only user sessions are sampled: anon requests aggregate
    without IP rows (DR-AL8); orbit has no api-key surface yet.
  - `auth_sessions` — login-session bookkeeping (DR-AL3, NOT enforcement):
    row on login, throttled last_seen stamps via the flush cycle, the
    expiry sweep turns the silent 12h cookie death into a real
    `auth.session_expired` audit event (DR-AL4).

  The request path only casts into bounded in-process buffers; a 15s timer
  flushes (swap-then-write: a failed flush loses at most one interval), a
  5-min timer expires, a daily timer prunes with batched oldest-first
  deletes (gap-lock rule), retention days from the settings registry.
  """

  use GenServer

  require Logger

  @flush_ms 15_000
  @expire_ms 5 * 60_000
  @prune_ms 24 * 60 * 60_000

  @events_per_flush 50
  @last_seen_throttle_ms 60_000
  @prune_batch 10_000

  # Keep in sync with the python SessionMiddleware max_age (12h) — the
  # registry sweep must agree with when the cookie actually dies.
  @session_max_age_s 12 * 60 * 60

  def start_link(opts) do
    GenServer.start_link(__MODULE__, :ok, name: Keyword.get(opts, :name, __MODULE__))
  end

  @doc "Fresh 32-hex session id (uuid4-hex parity)."
  def new_sid, do: Base.encode16(:crypto.strong_rand_bytes(16), case: :lower)

  @doc "Count one request (bounded buffers; safe without the server: drop)."
  def record_request(ptype, pkey, ip, method, path, status, opts \\ []) do
    # No double-default head: a 7-arg call once bound server="user" and
    # 500'd every page via before_send (hub.ex:92 lesson, relearned live).
    record_request_on(__MODULE__, ptype, pkey, ip, method, path, status, opts)
  end

  @doc false
  def record_request_on(server, ptype, pkey, ip, method, path, status, opts) do
    GenServer.cast(server, {:record, ptype, pkey, ip, method, path, status, Map.new(opts)})
  catch
    :exit, _ -> :ok
  end

  @doc "Registry row for a freshly minted login session (synchronous)."
  def open_session(server \\ __MODULE__, sid, user_id, ip) do
    GenServer.call(server, {:open_session, sid, user_id, ip})
  catch
    :exit, _ -> :ok
  end

  @doc "Mark a registry row ended (logout). Unknown/missing sid: no-op."
  def close_session(server \\ __MODULE__, sid, reason)
  def close_session(_server, nil, _reason), do: :ok

  def close_session(server, sid, reason) do
    GenServer.call(server, {:close_session, sid, reason})
  catch
    :exit, _ -> :ok
  end

  @doc "Synchronous flush (test/ops seam); returns the aggregated count."
  def flush(server \\ __MODULE__), do: GenServer.call(server, :flush)

  # -- pure buffer transitions (unit-tested) --------------------------------

  @doc false
  def empty_buffers do
    %{agg: %{}, last_ip: %{}, events: [], seen: %{}, stamped: %{}}
  end

  @doc false
  def add_request(buffers, ptype, pkey, ip, method, path, status, opts, now, mono_ms) do
    buffers
    |> bump_agg(ptype, pkey, ip)
    |> sample_event(ptype, ip, method, path, status, opts[:user_id], now)
    |> stamp_seen(opts[:sid], now, mono_ms)
  end

  defp bump_agg(buffers, ptype, pkey, ip) do
    agg = Map.update(buffers.agg, {ptype, pkey}, 1, &(&1 + 1))

    last_ip =
      if ip in [nil, ""] do
        buffers.last_ip
      else
        Map.put(buffers.last_ip, {ptype, pkey}, String.slice(ip, 0, 45))
      end

    %{buffers | agg: agg, last_ip: last_ip}
  end

  # Only user sessions are sampled — anon aggregates without IP rows
  # (DR-AL8: data minimisation; forensics detail stays in the docker log).
  defp sample_event(buffers, "user", ip, method, path, status, user_id, now) do
    event = %{
      ts: now,
      user_id: user_id,
      ip: String.slice(ip || "?", 0, 45),
      method: String.slice(method, 0, 8),
      path: String.slice(path, 0, 255),
      status: status
    }

    %{buffers | events: Enum.take([event | buffers.events], @events_per_flush)}
  end

  defp sample_event(buffers, _ptype, _ip, _method, _path, _status, _user_id, _now), do: buffers

  defp stamp_seen(buffers, nil, _now, _mono_ms), do: buffers

  defp stamp_seen(buffers, sid, now, mono_ms) do
    case buffers.stamped[sid] do
      last when is_integer(last) and mono_ms - last < @last_seen_throttle_ms ->
        buffers

      _ ->
        %{
          buffers
          | stamped: Map.put(buffers.stamped, sid, mono_ms),
            seen: Map.put(buffers.seen, sid, now)
        }
    end
  end

  # -- GenServer ------------------------------------------------------------

  @impl true
  def init(:ok) do
    Process.send_after(self(), :flush, @flush_ms)
    Process.send_after(self(), :expire, @expire_ms)
    Process.send_after(self(), :prune, @prune_ms)
    {:ok, empty_buffers()}
  end

  @impl true
  def handle_cast({:record, ptype, pkey, ip, method, path, status, opts}, buffers) do
    now = DateTime.utc_now()
    mono = System.monotonic_time(:millisecond)
    {:noreply, add_request(buffers, ptype, pkey, ip, method, path, status, opts, now, mono)}
  end

  @impl true
  def handle_call({:open_session, sid, user_id, ip}, _from, buffers) do
    now = DateTime.utc_now() |> DateTime.to_naive()

    Orbit.Repo.query!(
      "INSERT INTO auth_sessions (sid, user_id, ip, created_at, last_seen_at) " <>
        "VALUES (?, ?, ?, ?, ?)",
      [sid, user_id, ip && String.slice(ip, 0, 45), now, now]
    )

    stamped = Map.put(buffers.stamped, sid, System.monotonic_time(:millisecond))
    {:reply, :ok, %{buffers | stamped: stamped}}
  rescue
    error ->
      Logger.warning("access.open_session_failed error=#{Exception.message(error)}")
      {:reply, :error, buffers}
  end

  def handle_call({:close_session, sid, reason}, _from, buffers) do
    Orbit.Repo.query!(
      "UPDATE auth_sessions SET ended_at = ?, end_reason = ? " <>
        "WHERE sid = ? AND ended_at IS NULL",
      [DateTime.utc_now() |> DateTime.to_naive(), reason, sid]
    )

    buffers = %{
      buffers
      | stamped: Map.delete(buffers.stamped, sid),
        seen: Map.delete(buffers.seen, sid)
    }

    {:reply, :ok, buffers}
  rescue
    error ->
      Logger.warning("access.close_session_failed error=#{Exception.message(error)}")
      {:reply, :error, buffers}
  end

  def handle_call(:flush, _from, buffers) do
    {n, rest} = do_flush(buffers)
    {:reply, n, rest}
  end

  @impl true
  def handle_info(:flush, buffers) do
    Process.send_after(self(), :flush, @flush_ms)
    {_n, rest} = do_flush(buffers)
    {:noreply, rest}
  end

  def handle_info(:expire, buffers) do
    Process.send_after(self(), :expire, @expire_ms)
    expire_sessions()
    {:noreply, buffers}
  end

  def handle_info(:prune, buffers) do
    Process.send_after(self(), :prune, @prune_ms)
    prune()
    {:noreply, buffers}
  end

  # Swap-then-write (throttle stamps survive — they are bookkeeping, not data).
  defp do_flush(%{agg: agg, events: events, seen: seen} = buffers)
       when agg == %{} and events == [] and seen == %{} do
    {0, buffers}
  end

  defp do_flush(buffers) do
    %{agg: agg, last_ip: last_ip, events: events, seen: seen} = buffers
    rest = %{buffers | agg: %{}, last_ip: %{}, events: [], seen: %{}}

    now = DateTime.utc_now() |> DateTime.to_naive()
    bucket = %{now | minute: 0, second: 0, microsecond: {0, 0}}

    for {{ptype, pkey}, n} <- agg do
      Orbit.Repo.query!(
        "INSERT INTO access_stats (bucket, principal_type, principal_key, count, last_ip) " <>
          "VALUES (?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE count = count + ?, " <>
          "last_ip = COALESCE(?, last_ip)",
        [bucket, ptype, pkey, n, last_ip[{ptype, pkey}], n, last_ip[{ptype, pkey}]]
      )
    end

    for e <- Enum.reverse(events) do
      Orbit.Repo.query!(
        "INSERT INTO access_events (ts, user_id, ip, method, path, status) " <>
          "VALUES (?, ?, ?, ?, ?, ?)",
        [DateTime.to_naive(e.ts), e.user_id, e.ip, e.method, e.path, e.status]
      )
    end

    for {sid, ts} <- seen do
      Orbit.Repo.query!(
        "UPDATE auth_sessions SET last_seen_at = ? WHERE sid = ? AND ended_at IS NULL",
        [DateTime.to_naive(ts), sid]
      )
    end

    n = agg |> Map.values() |> Enum.sum()
    if n > 0, do: Logger.debug("access.flushed requests=#{n}")
    {n, rest}
  rescue
    error ->
      Logger.warning("access.flush_failed error=#{Exception.message(error)}")
      {0, %{buffers | agg: %{}, last_ip: %{}, events: [], seen: %{}}}
  end

  # Bookkeeping, not enforcement (DR-AL3/AL4): the cookie died on its own at
  # max_age — here the silent auto-logout becomes a visible audit event.
  defp expire_sessions do
    cutoff =
      DateTime.utc_now() |> DateTime.add(-@session_max_age_s, :second) |> DateTime.to_naive()

    %{rows: rows} =
      Orbit.Repo.query!(
        "SELECT sid, user_id, ip FROM auth_sessions " <>
          "WHERE ended_at IS NULL AND created_at < ?",
        [cutoff]
      )

    now = DateTime.utc_now() |> DateTime.to_naive()

    for [sid, user_id, ip] <- rows do
      Orbit.Repo.query!(
        "UPDATE auth_sessions SET ended_at = ?, end_reason = 'expired' WHERE sid = ?",
        [now, sid]
      )

      Orbit.Audit.write(
        action: "auth.session_expired",
        result: "ok",
        user_id: user_id,
        source_ip: ip
      )
    end

    if rows != [], do: Logger.info("access.sessions_expired sessions=#{length(rows)}")
    :ok
  rescue
    error -> Logger.warning("access.expire_failed error=#{Exception.message(error)}")
  end

  defp prune do
    events_days = Orbit.Settings.effective("access_events_retention_days")
    stats_days = Orbit.Settings.effective("access_stats_retention_days")
    sessions_days = Orbit.Settings.effective("access_sessions_retention_days")

    now = DateTime.utc_now()
    prune_events_batched(now |> DateTime.add(-events_days, :day) |> DateTime.to_naive())

    Orbit.Repo.query!(
      "DELETE FROM access_stats WHERE bucket < ?",
      [now |> DateTime.add(-stats_days, :day) |> DateTime.to_naive()]
    )

    Orbit.Repo.query!(
      "DELETE FROM auth_sessions WHERE ended_at IS NOT NULL AND ended_at < ?",
      [now |> DateTime.add(-sessions_days, :day) |> DateTime.to_naive()]
    )

    :ok
  rescue
    error -> Logger.warning("access.prune_failed error=#{Exception.message(error)}")
  end

  defp prune_events_batched(cutoff) do
    %{num_rows: n} =
      Orbit.Repo.query!(
        "DELETE FROM access_events WHERE ts < ? ORDER BY ts LIMIT ?",
        [cutoff, @prune_batch]
      )

    if n >= @prune_batch, do: prune_events_batched(cutoff)
  end
end
